### PR TITLE
*: add keyvisual service

### PR DESCRIPTION
**Pull request by nolouch**
Signed-off-by: nolouch <nolouch@gmail.com>

### What problem does this PR solve? 
Introduce the key visual service.

### What is changed and how it works?
add an empty service now. 

### Check List 

Tests 

 - Unit test
 - Integration test

### CODE REVIEW

File: pkg/apiutil/serverapi/middleware.go [hunk 1]
```diff
@@ -40,32 +41,76 @@ const (
...
+		DisableKeepAlives: true,
+	},
+}
+
+type runtimeServiceAuth struct {
+	s *server.Server
+}
+
+// NewRuntimeServiceAuth checks if the path is invalid.
+func NewRuntimeServiceAuth(s *server.Server) *runtimeServiceAuth {
```
Comment (HunDunDM) [inline pkg/apiutil/serverapi/middleware.go]:
```suggestion
func NewRuntimeServiceAuth(s *server.Server) Handler {
```
Exported function with unexported return type

File: pkg/apiutil/serverapi/middleware.go [hunk 2]
```diff
@@ -40,32 +41,76 @@ const (
...
+	if IsServiceAllowed(h.s, r.RequestURI) {
+		next(w, r)
+		return
+	}
+
+	http.Error(w, "no service", http.StatusServiceUnavailable)
+}
+
+// IsServiceAllowed checks the service through the path.
+func IsServiceAllowed(s *server.Server, path string) bool {
```
Comment (HunDunDM) [inline pkg/apiutil/serverapi/middleware.go]: This function is called after routing, and its return value should be unchanged after startup, and can be determined at NewRuntimeServiceAuth. Can the configuration change at runtime?

In addition, assuming that there is `foo` in RuntimeServices, a `/pd/apis/keyvisual/v1/foo` like path can maliciously bypass the check. Can we use APIGroup.Name directly to check.

File: pkg/apiutil/serverapi/middleware.go [hunk 3]
```diff
@@ -40,32 +41,76 @@ const (
...
-func newRedirector(s *server.Server) *redirector {
+// NewRedirector redirects request to the leader if needs to be handled in the leader.
+func NewRedirector(s *server.Server) *redirector {
 	return &redirector{s: s}
 }
 
 func (h *redirector) ServeHTTP(w http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
-	allowFollowerHandle := len(r.Header.Get(allowFollowerHandle)) > 0
+	allowFollowerHandle := len(r.Header.Get(AllowFollowerHandle)) > 0
 	if !h.s.IsClosed() && (h.s.GetMember().IsLeader() || allowFollowerHandle) {
```
Comment (HunDunDM) [inline pkg/apiutil/serverapi/middleware.go]:
```suggestion
	if !h.s.IsClosed() && (allowFollowerHandle || h.s.GetMember().IsLeader()) {
```

Will short circuit operations be faster?

**Comment by HunDunDM** [comment]
When key visual is configured not to run, the background thread will still run?

File: pkg/apiutil/serverapi/middleware.go (hunk 2 above)
Comment (nolouch) [reply pkg/apiutil/serverapi/middleware.go]: Yes, Server option can be changed in the runtime. `/pd/apis/keyvisual/v1/foo` can not bypass, what do you mean?

**Comment by nolouch** [comment]
@HunDunDM  Yes, only one idle goroutine.

File: server/config/config.go [hunk 4]
```diff
@@ -199,6 +199,8 @@ const (
 	defaultDisableErrorVerbose = true
 )
 
+var defaultRuntimeServices = []string{""}
```
Comment (disksing) [inline server/config/config.go]: why not use `[]string{}`?

File: pkg/apiutil/serverapi/middleware.go [hunk 5]
```diff
@@ -40,32 +42,77 @@ const (
...
+	opt := s.GetServerOption()
+	cfg := opt.LoadPDServerConfig()
+	if cfg != nil {
+		for _, allow := range cfg.RuntimeServices {
+			if len(allow) != 0 && group.Name == allow {
+				return true
+			}
+		}
+	}
+	// for core path
```
Comment (HunDunDM) [inline pkg/apiutil/serverapi/middleware.go]: First determine if it is a core, is this better?

File: pkg/apiutil/serverapi/middleware.go [hunk 6]
```diff
@@ -40,32 +42,77 @@ const (
...
+	http.Error(w, "no service", http.StatusServiceUnavailable)
+}
+
+// IsServiceAllowed checks the service through the path.
+func IsServiceAllowed(s *server.Server, group server.APIGroup) bool {
+	opt := s.GetServerOption()
+	cfg := opt.LoadPDServerConfig()
+	if cfg != nil {
+		for _, allow := range cfg.RuntimeServices {
+			if len(allow) != 0 && group.Name == allow {
```
Comment (disksing) [inline pkg/apiutil/serverapi/middleware.go]: `allow != ""`? I think we don't need it if `RuntimeService` initialized as `[]string{}`.

File: pkg/keyvisual/service.go [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+	endTimeString := form.Get("endtime")
+	typ := form.Get("type")
+	endTime := time.Now()
+	startTime := endTime.Add(-1200 * time.Minute)
+
+	if startTimeString != "" {
+		tsSec, err := strconv.ParseInt(startTimeString, 10, 64)
+		if err != nil {
+			log.Error("parse ts failed", zap.Error(err))
+
```
Comment (disksing) [inline pkg/keyvisual/service.go]: empty line?

File: pkg/keyvisual/service.go [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+		}
+		startTime = time.Unix(tsSec, 0)
```
Comment (disksing) [inline pkg/keyvisual/service.go]: should use default value when parse fail?

File: pkg/keyvisual/service.go [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+	}
+	if endTimeString != "" {
+		tsSec, err := strconv.ParseInt(endTimeString, 10, 64)
+		if err != nil {
+			log.Error("parse ts failed", zap.Error(err))
+
```
Comment (disksing) [inline pkg/keyvisual/service.go]: empty line

File: pkg/keyvisual/service.go [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+		}
+		endTime = time.Unix(tsSec, 0)
```
Comment (disksing) [inline pkg/keyvisual/service.go]: should use default value when parse fail?

File: pkg/keyvisual/service.go [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+
```
Comment (disksing) [inline pkg/keyvisual/service.go]: remove empty line

File: pkg/apiutil/serverapi/middleware.go [hunk 12]
```diff
@@ -40,32 +42,77 @@ const (
...
+	},
+}
+
+type runtimeServiceAuth struct {
+	s     *server.Server
+	group server.APIGroup
+}
+
+// NewRuntimeServiceAuth checks if the path is invalid.
+func NewRuntimeServiceAuth(s *server.Server, group server.APIGroup) negroni.Handler {
```
[the remark this answers is not in the document]
Comment (HunDunDM) [reply pkg/apiutil/serverapi/middleware.go]: How about `NewRuntimeServiceEntrance`

File: server/config/config.go (hunk 4 above)
Comment (HunDunDM) [reply server/config/config.go]: Maybe `var defaultRuntimeServices []string` is more recommended.

File: pkg/keyvisual/service.go (hunk 8 above)
Comment (HunDunDM) [reply pkg/keyvisual/service.go]: I think it should return 403.

File: pkg/apiutil/serverapi/middleware.go [hunk 13]
```diff
@@ -40,32 +42,77 @@ const (
...
+	cfg := opt.LoadPDServerConfig()
+	if cfg != nil {
+		for _, allow := range cfg.RuntimeServices {
+			if len(allow) != 0 && group.Name == allow {
+				return true
+			}
+		}
+	}
+	// for core path
+	if group.IsCore {
```
[the remark this answers is not in the document]
Comment (HunDunDM) [reply pkg/apiutil/serverapi/middleware.go]:
```go
	// for core path
	if group.IsCore {
		return true
	}
	
	opt := s.GetServerOption()
	cfg := opt.LoadPDServerConfig()
	if cfg != nil {
		for _, allow := range cfg.RuntimeServices {
			if len(allow) != 0 && group.Name == allow {
				return true
			}
		}
	}

	return false
```

File: server/config/config.go (hunk 4 above)
Comment (disksing) [reply server/config/config.go]: em, that's is different. `var defaultRuntimeServices []string` creates a `nil` variable, which will marshal to `null`

File: pkg/keyvisual/service.go (hunk 8 above)
Comment (disksing) [reply pkg/keyvisual/service.go]: > I think it should return 403.

Good idea.

File: server/config/config.go (hunk 4 above)
Comment (HunDunDM) [reply server/config/config.go]: get

File: pkg/apiutil/serverapi/middleware.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (nolouch) [reply pkg/apiutil/serverapi/middleware.go]: I think validator better. :)

**Pull request merged**